docs/blog/2021-07-18-macro.md

```diff
@@ -0,0 +1,105 @@
+
+I also found that `proc_macro_derive`s have no way to get the `#[derive(..)]` attribute of the type, and I wanted to also bind derived trait implementations (at least for `Default`, because some types have no other constructors). So I removed the derive and moved the implementation to the `#[bindlang]` attribute macro, which now works on types, impl blocks and fns.
+
+**Beware: This makes use of the (unspecified, but as of now working) top-down order of macro expansion to work!**
+
+## Dirty tricks avoided
+
+There is a `Span::mixed_context()` variant that will yield semi-hygienic macros (like with `macro_rules`). However, this looked risky (macro hygiene is there to protect us, so we better have a good reason to override it), so I took the data oriented approach, collecting the info I needed to create the code in the `lazy_static` to walk within `bindlang_main!()`. I still tried to generate the trait impls for marshalling directly in the attribute macro, but this again ran into macro hygiene trouble, because I could not recreate the virtual dispatch table identifiers. After moving this to the main macro, too, the macro finally expanded successfully.
+
+Except it didn’t compile successfully.
+
+I had forgotten to `use` the items I was creating code for in the macro, and koto requires all external types to implement `Display`. So I added those imports as macro arguments and added the `Display` `impl`s to be met with a type inference error within the macro invocation. Clearly I needed some type annotations, but the error message only showed me the macro invocation, which was pretty unhelpful.
+
+## Expanding Our Vision
+
+My solution to debug those was to `cargo expand` the code, comment out the macro invocation and copy the expanded portions in its place so that my IDE would pinpoint the error for me. I had to manually un-expand some `format!` invocations so the code would resolve correctly, and finally found where I needed more type annotations. With those added, the code finally compiled. Whew!
+
+I then extended the bindings to also cover trait impls and `Option`s, while my colleague Christos changed the code to marshall Rust values into koto values to mangle `Result::Err(_)` into koto’s runtime errors. Finally, remembering that implicit constructors (structs and enum variants) are also useful, I added support to binding those if public. There was another error where intermediate code generated wouldn't parse, but some `eprintln!` debugging helped pinpoint the piece of code where it happened.
```

Comment:
It would be nice to add a link to the source code for your more curious readers

Reply:
I'll add the link once the code is merged. Otherwise the link is going to get stale quickly.